src/ray/common/scheduling/cluster_resource_scheduler.cc

```diff
@@ -487,3 +621,225 @@ std::string ClusterResourceScheduler::DebugString(void) {
+          (UnitInstanceResources.find(i) != UnitInstanceResources.end()),
+          &local_resources_.predefined_resources[i]);
+    }
+  }
+
+  if (node_resources.custom_resources.size() == 0) {
+    return;
+  }
+
+  for (auto it = node_resources.custom_resources.begin();
+       it != node_resources.custom_resources.end(); ++it) {
+    if (it->second.total > 0) {
+      ResourceInstanceCapacities instance_list;
+      InitResourceInstances(it->second.total, false, &instance_list);
+      local_resources_.custom_resources.emplace(it->first, instance_list);
+    }
+  }
+}
+
+void ClusterResourceScheduler::AddAvailableResourceInstances(
+    std::vector<double> available,
+    ResourceInstanceCapacities *resource_instances /* return */) {
+  for (size_t i = 0; i < available.size(); i++) {
+    resource_instances->available[i] = std::min(
+        resource_instances->available[i] + available[i], resource_instances->total[i]);
+  }
+}
+
+void ClusterResourceScheduler::SubtractAvailableResourceInstances(
+    std::vector<double> available,
+    ResourceInstanceCapacities *resource_instances /* return */) {
+  RAY_CHECK(available.size() == resource_instances->available.size());
+
+  for (size_t i = 0; i < available.size(); i++) {
+    resource_instances->available[i] =
+        std::max(resource_instances->available[i] - available[i], 0.);
+  }
+}
+
+bool ClusterResourceScheduler::AllocateResourceInstances(
+    double demand, bool soft, std::vector<double> &available,
+    std::vector<double> *allocation /* return */) {
+  allocation->resize(available.size());
+
+  if (available.size() == 1) {
+    // This resource has just an instance.
+    if (available[0] >= demand) {
+      available[0] -= demand;
+      (*allocation)[0] = demand;
+      return true;
+    } else {
+      if (soft) {
+        available[0] = 0;
+        return true;
+      }
+      // Not enough capacity.
+      return false;
+    }
+  }
+
+  // If resources has multiple instances, each instance has total capcity of 1.
+  // If demand is greater than 1., allocate full unit-capacity instances until the
+  // remaining demand is fractional.
+  if (demand >= 1.) {
+    for (size_t i = 0; i < available.size(); i++) {
+      if (available[i] == 1.) {
+        // Allocate a full unit-capacity instance.
+        (*allocation)[i] = 1.;
+        available[i] = 0;
+        demand -= 1.;
+      }
+      if (demand < 1.) {
+        break;
+      }
+    }
+  }
+
+  if (soft) {
+    // Just get as much resources as available.
+    for (size_t i = 0; i < available.size(); i++) {
+      if (available[i] >= demand) {
+        available[i] -= demand;
+        (*allocation)[i] = demand;
+        return true;
+      } else {
+        (*allocation)[i] = available[i];
+        demand -= available[i];
+        available[i] = 0;
+      }
+    }
+    return true;
+  }
+
+  if (demand >= 1.) {
+    // Cannot satisfy a deman greater than one if no unit caapcity resource is available.
```

Comment:
typos

Reply:
Seems like this error check should be higher up in the function, before we start modifying `demand`

Reply:
Also, in general it's preferable to not modify the arguments to the function. Better to make a separate variable (e.g., `remaining_demand` so it's clear when you're checking against the passed argument vs. current state of the function.

Reply:
The intention that unless the resources are soft we only allocate at most one fractional resource. For instance. If we ask for 2.5 CPUs, we get two full CPUs and 0.5 of another CPU, not 0.25 of one CPU and another 0.25 of another CPU. I'll make that clear.

Regarding remaining_demand, that sounds good.

Reply:
Are you planning to modify this to use `remaining_demand`?

Reply:
Already did it!